eventhubs/data-plane/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/SendLinkHandler.java

```diff
@@ -49,10 +48,10 @@ public void onLinkRemoteOpen(Event event) {
                     TRACE_LOGGER.info(String.format(Locale.US, "onLinkRemoteOpen linkName[%s], remoteTarget[%s]", sender.getName(), link.getRemoteTarget()));
                 }
 
-                synchronized (this.firstFlow) {
-                    this.isFirstFlow = false;
+                if (this.isFirstFlow.compareAndSet(true, false)) {
```

Comment:
I'm not sure this change is the behaviour you want.  
The code to the left __always__ goes into this critical section and sets it to false.
This will _only_ return true if the expected value was equal to `true` and then you set it to `false`. So, if at the moment this line is hit, if the expected value of `true` does not match the current value (which would be false), then it won't go into the if statement.

Reply:
In addition, its possible for multiple threads to, at the same time call `this.msgSender.onOpenComplete(null)` because this isn't inside inside the same synchronized block.

Reply:
There is a discussion on this change on Teams Java Channel at 3/26/ 5:07PM.  Just copy of portion of conversation to here: "(the original author of EventHub client lib) regarding replacing "[Object firstFlow, boolean isFirstFlow, synchronized]" with "[AtomicBoolean isFirstFlow, compareAndSet]", he seems fine with it."

Reply:
Ahh gotcha. :) 👍